addons/binding/org.openhab.binding.plclogo/src/main/java/org/openhab/binding/plclogo/handler/PLCPulseHandler.java

```diff
@@ -0,0 +1,338 @@
+        }
+    }
+
+    @Override
+    public void setData(final byte[] data) {
+        if (!isThingOnline()) {
+            return;
+        }
+
+        if (data.length != getBufferLength()) {
+            logger.warn("Received and configured data sizes does not match.");
+            return;
+        }
+
+        final List<Channel> channels = thing.getChannels();
+        if (channels.size() != getNumberOfChannels()) {
+            logger.warn("Received and configured channels sizes does not match.");
+            return;
+        }
+
+        for (final Channel channel : channels) {
+            final ChannelUID channelUID = channel.getUID();
+            final String name = getBlockFromChannel(channel);
+
+            final int bit = getBit(name);
+            final int address = getAddress(name);
+            if ((address != INVALID) && (bit != INVALID)) {
+                final DecimalType state = (DecimalType) getOldValue(channelUID.getId());
+                if (STATE_CHANNEL.equalsIgnoreCase(channelUID.getId())) {
+                    boolean value = S7.GetBitAt(data, address - getBase(name), bit);
+                    if ((state == null) || ((value ? 1 : 0) != state.intValue())) {
+                        updateChannel(channel, value);
+                    }
+                } else if (OBSERVE_CHANNEL.equalsIgnoreCase(channelUID.getId())) {
+                    handleCommand(channelUID, RefreshType.REFRESH);
+                    if ((state != null) && !state.equals(getOldValue(channelUID.getId()))) {
+                        final Integer pulse = config.getPulseLength();
+                        scheduler.schedule(new Runnable() {
+                            private final String name = config.getBlockName();
+                            private final PLCLogoClient client = getLogoClient();
+
+                            @Override
+                            public void run() {
+                                final byte[] data = new byte[1];
+                                S7.SetBitAt(data, 0, 0, state.intValue() == 1 ? true : false);
+                                final int address = 8 * getAddress(name) + getBit(name);
+                                int result = client.writeDBArea(1, address, data.length, S7Client.S7WLBit, data);
+                                if (result == 0) {
+                                    setOldValue(channelUID.getId(), null);
+                                } else {
+                                    logger.warn("Can not write data to LOGO!: {}.", S7Client.ErrorText(result));
+                                }
+                            }
+                        }, pulse.longValue(), TimeUnit.MILLISECONDS);
+                    }
+                } else {
+                    logger.warn("Invalid channel {} found.", channelUID);
+                }
+            } else {
+                logger.warn("Invalid channel {} found.", channelUID);
+            }
+        }
+
+        /*
```

Comment:
This is the third time I see such a block, is this code maybe to repetitive? Can you check the rest.